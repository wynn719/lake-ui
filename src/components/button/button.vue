<template>
  <button type="button" class="lake-btn" :class="btnClass" :disabled="disabled" @click="onClick">
    <i class="lake-loading" v-if="loading"></i>
    <!-- @slot 自定义按钮的图标 -->
    <slot name="icon"></slot>
    <!-- @slot 自定义按钮里的内容 -->
    <slot></slot>
  </button>
</template>

<script>
const BTN_TYPES = ['default', 'primary', 'warning', 'outline'];
const BTN_SIZES = ['small', 'normal', 'large'];

/**
 * 按钮用于开始一个即时操作。
 */
export default {
  name: 'lake-button',
  props: {
    /**
     * 按钮失效状态
     */
    disabled: {
      type: Boolean,
      default: false,
    },
    /**
     * 设置按钮类型，可选值为 ['default', 'primary', 'warning', 'outline'] 或者不设
     */
    type: {
      type: String,
      default: 'default',
      validator(type) {
        return type === '' || BTN_TYPES.includes(type);
      },
    },
    /**
     * 设置按钮大小，可选值为 ['small', 'normal', 'large'] 或者不设
     */
    size: {
      type: String,
      default: '',
      validator(size) {
        return size === '' || BTN_SIZES.includes(size);
      },
    },
    /**
     * 按钮是否为行内元素
     */
    inline: {
      type: Boolean,
      default: false,
    },
    /**
     * 设置按钮载入状态
     */
    loading: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    btnClass() {
      return [
        this.inline ? 'lake-btn-inline' : '',
        this.type ? `lake-btn-${this.type}` : '',
        this.size ? `lake-btn-${this.size}` : '',
        this.loading ? `lake-btn-loading` : '',
      ];
    },
  },
  methods: {
    /**
     * 手动触发按钮的点击事件
     *
     * @param {object} $event
     * @public
     */
    onClick($event) {
      if (this.disabled) {
        $event.preventDefault();
        $event.stopPropagation();
        return;
      }

      /**
       * 点击按钮时的回调
       *
       * @event click
       * @type {event}
       * @property {DOMEvent} $event
       */
      this.$emit('click', $event);
    },
  },
};
</script>

<style lang="less">
@import '../../style/themes/default.less';
@import '../../style/common/mixins.less';
@btnPrefixClass: lake-btn;
/**
 * get backgroundColor active status and disabled status
 */
.btn-mutate(@backgroundColor) {
  @level: 7;

  &.@{btnPrefixClass}-loading,
  &:active {
    color: rgba(255, 255, 255, 0.6);
    background-color: .hsv-color(@backgroundColor, @level) [ @result];
    border-color: .hsv-color(@backgroundColor, @level) [ @result];
  }
  &:disabled {
    color: #fff;
    background-color: #d8dee6;
    border-color: #d8dee6;
  }
}
.btn-mutate-white(@backgroundColor) {
  &.@{btnPrefixClass}-loading,
  &:active {
    color: rgba(51, 51, 51, 0.6);
    background-color: #f5f7fa;
    border-color: #d8dee6;
  }
  &:disabled {
    color: rgba(51, 51, 51, 0.3);
    background-color: #ffffff;
    border-color: #d8dee6;
  }
}

.@{btnPrefixClass} {
  display: block;
  outline: 0 none;
  box-sizing: border-box;
  padding: 13px 15px;
  text-align: center;
  font-size: 16px;
  width: 100%;
  min-width: 64px;
  height: 48px;
  line-height: 22px;
  overflow: hidden;
  text-overflow: ellipsis;
  word-break: break-word;
  white-space: nowrap;
  color: #333;
  background-color: #fff;
  border: 1px solid #d8dee6;
  border-radius: 4px;
  user-select: none;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  -webkit-appearance: none;
  -webkit-font-smoothing: antialiased;
  .btn-mutate-white(#fff);

  &-primary {
    color: #fff;
    background-color: @brand-primary;
    border-color: @brand-primary;
    .btn-mutate(@brand-primary);
  }
  &-warning {
    color: #fff;
    background-color: @brand-error;
    border-color: @brand-error;
    .btn-mutate(@brand-error);
  }
  &-outline {
    color: @brand-primary;
    background-color: #fff;
    border-color: #008dfc;
  }
  &-small {
    font-size: 14px;
    height: 32px;
    line-height: 20px;
    padding: 5px 10px;
  }
  &-normal {
    font-size: 16px;
    height: 40px;
    line-height: 22px;
    padding: 8px 15px;
  }

  &-inline {
    display: inline-block;
    vertical-align: top;
    width: auto;
  }
  &-loading {
    display: flex;
    align-items: center;
    justify-content: center;
  }
  .lake-loading {
    display: block;
    width: 12px;
    height: 12px;
    margin-right: 6px;
    border: 2px solid #e8e8e8;
    border-top-color: #c5c5c5;
    border-radius: 100%;
    animation: rotation 0.5s linear 0s infinite;
  }

  @keyframes rotation {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(359deg);
    }
  }
}
</style>
