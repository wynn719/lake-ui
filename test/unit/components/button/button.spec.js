import { mount } from '@vue/test-utils';
import button from '@/components/button';

describe('button', () => {
  // basic test
  test('should render component and match snapshot', () => {
    const wrapper = mount(button);

    expect(wrapper.element).toMatchSnapshot();
  });

  // props test
  test('should have lake-btn-primary class when using type props', () => {
    const wrapper = mount(button, {
      propsData: {
        type: 'primary',
      },
    });

    expect(wrapper.classes('lake-btn-primary')).toBe(true);
  });

  test('should have lake-btn-small class when using size props', () => {
    const wrapper = mount(button, {
      propsData: {
        size: 'small',
      },
    });

    expect(wrapper.classes('lake-btn-small')).toBe(true);
  });

  test('should have lake-btn-inline class when using inline props', () => {
    const wrapper = mount(button, {
      propsData: {
        inline: true,
      },
    });

    expect(wrapper.classes('lake-btn-inline')).toBe(true);
  });

  test('should disabled click when using disabled props', () => {
    const wrapper = mount(button, {
      propsData: {
        disabled: true,
      },
    });
    const click = jest.fn();

    wrapper.vm.$on('click', click);
    wrapper.trigger('click');

    expect(click.mock.calls.length).toBe(0);
  });

  test('should render component with loading and match snapshot', () => {
    const wrapper = mount(button, {
      propsData: {
        loading: true,
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  // event test
  test('should emit a click event', () => {
    const wrapper = mount(button);
    const click = jest.fn();

    wrapper.vm.$on('click', click);
    wrapper.trigger('click');

    expect(click.mock.calls.length).toBe(1);
  });

  // slot test
  test('should render component with slot and match snapshot', () => {
    const wrapper = mount(button, {
      slots: {
        default: 'submit',
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
