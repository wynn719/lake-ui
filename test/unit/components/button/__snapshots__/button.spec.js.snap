// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button should render component and match snapshot 1`] = `
<button
  class="lake-btn lake-btn-default"
  type="button"
>
  <!---->
    
</button>
`;

exports[`button should render component with loading and match snapshot 1`] = `
<button
  class="lake-btn lake-btn-default lake-btn-loading"
  type="button"
>
  <i
    class="lake-loading"
  />
    
</button>
`;

exports[`button should render component with slot and match snapshot 1`] = `
<button
  class="lake-btn lake-btn-default"
  type="button"
>
  <!---->
    submit
</button>
`;
